src/main/scala/ai/privado/languageEngine/csharp/tagger/collection/CollectionTagger.scala

```diff
@@ -0,0 +1,139 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * For more information, contact support@privado.ai
+ *
+ */
+
+package ai.privado.languageEngine.csharp.tagger.collection
+
+import ai.privado.cache.RuleCache
+import ai.privado.languageEngine.java.tagger.collection.CollectionUtility
+import ai.privado.model.{Constants, RuleInfo}
+import ai.privado.tagger.PrivadoParallelCpgPass
+import io.shiftleft.codepropertygraph.generated.Cpg
+import io.shiftleft.codepropertygraph.generated.nodes.{Annotation, Method}
+import io.shiftleft.semanticcpg.language.*
+import org.slf4j.LoggerFactory
+
+import scala.collection.mutable
+
+class CollectionTagger(cpg: Cpg, ruleCache: RuleCache) extends PrivadoParallelCpgPass[RuleInfo](cpg) {
+  protected val methodUrlMap: mutable.HashMap[Long, String] = mutable.HashMap[Long, String]()
+  protected val classUrlMap: mutable.HashMap[Long, String]  = mutable.HashMap[Long, String]()
+  private val logger                                        = LoggerFactory.getLogger(this.getClass)
+
+  override def generateParts(): Array[RuleInfo] =
+    ruleCache.getRule.collections.filter(_.catLevelTwo == Constants.annotations).toArray
+
+  override def runOnPart(builder: DiffGraphBuilder, ruleInfo: RuleInfo): Unit = {
+    classUrlMap.addAll(collectAnnotatedUrlsFromClasses(ruleInfo.combinedRulePattern))
+    val (methodUrls, collectionMethodsCache) = collectAnnotatedUrlsFromMethods(ruleInfo.combinedRulePattern)
+    methodUrlMap.addAll(methodUrls)
+    tagSources(builder, ruleInfo, collectionMethodsCache)
+  }
+
+  protected def tagSources(
+    builder: DiffGraphBuilder,
+    ruleInfo: RuleInfo,
+    collectionMethodsCache: List[Method]
+  ): Unit = {
+    CollectionUtility.tagDirectSources(
+      builder,
+      collectionMethodsCache,
+      ruleCache.getRule.sources,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+
+    CollectionUtility.tagDerivedSources(
+      cpg,
+      builder,
+      collectionMethodsCache,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+  }
+
+  private def getUrlFromAnnotationsCode(annotation: Annotation): String = {
+    // If we have [controller] in annotation code, then replace with the class name of controller
+    //
+    //    [Route("api/some/[controller]")]                      <-- /api/some/email
+    //    class EmailController {
+    //      [HttpGet("[controller]/[action]/{id}")]             <-- /api/some/email/edit/{id}
+    //      public IActionResult Edit(int id)
+    //      {
+    //        return ControllerContext.MyDisplayRouteInfo(id);
+    //      }
+    //    }
+    //
+    val controllerName = if annotation.code.contains("[controller]") then {
+      cpg.typeDecl
+        .where(_.annotation.codeExact(annotation.code))
+        .where(_.name(".*Controller"))
+        .name
```

Comment:
As we already have the `annotation` node, we can directly go to the respective typeDecl by `annotation.typeDecl`, instead `cpg.typeDecl`